### ISSUE TITLE
journalctl --list-boots hangs with 100% cpu usage.

### ISSUE THREAD

**Issue by fat-lobyte** [systemd/systemd#617]
Hi!

I wanted to list all the boots in the journal, so I ran `journalctl --list-boots`. The application does not finish and is permanently at 100 % cpu usage.

I have limited the disk usage to 2GB and `journalctl --disk-usage` reports

```
Archived and active journals take up 2.1G on disk.
```

I ran `journalctl --verify`, and came up with a few corrupt log files. I deleted those, but the problem still persists.

Does anyone have an idea what is going on?

**Comment by zonque** [comment]
What does `strace journalctl --list-boots` print?

**Comment by fat-lobyte** [comment]
I'm very sorry, I'm afraid I "fixed" the issue on my box with running `journalctl --vacuum-size=1G`
Therefor I cannot reproduce it anymore. If I run into it again, I will post the results here.

**Comment by lnykryn** [comment]
I think we have the same issue in systemd in rhel.
Here is a strace from the customer
https://people.redhat.com/lnykryn/60-journalctl_strace_from_start.txt.bz2

If you don't think that it is relevant, feel free to close this issue again.

**Comment by Celti** [comment]
Occurs for me as well on systemd-229, Arch Linux.

An strace is available at https://ilmatar.celti.name/~celti/esoterica/journalctl-list-boots-hang-strace.txt.xz — be warned that that's a 29KB xz that inflates to 105MB of text. 99% of it is an endless cycle of mmap() and munmap():

```
mmap(NULL, 8388608, PROT_READ, MAP_SHARED, 72, 0xfd4000) = 0x7f588d8d4000
munmap(0x7f588e0d4000, 8388608)         = 0
mmap(NULL, 8388608, PROT_READ, MAP_SHARED, 74, 0x462000) = 0x7f588e0d4000
munmap(0x7f588ffb2000, 8388608)         = 0
mmap(NULL, 8388608, PROT_READ, MAP_SHARED, 75, 0x161d000) = 0x7f588ffb2000
munmap(0x7f588d8d4000, 8388608)         = 0
mmap(NULL, 8388608, PROT_READ, MAP_SHARED, 75, 0x1f4e000) = 0x7f588d8d4000
munmap(0x7f588f7b2000, 8388608)         = 0
mmap(NULL, 7200768, PROT_READ, MAP_SHARED, 77, 0x922000) = 0x7f588f8d4000
munmap(0x7f588e0d4000, 8388608)         = 0
```

**Comment by Celti** [comment]
After doing `journalctl --vacuum-time="5 months"` and pruning about 1GB of archived journals (backed up all of /var/log/journal first so that I can restore it for purposes of bug-hunting), I tried again. This time it hangs with an endlessly incrementing series of calls to brk().

strace at https://ilmatar.celti.name/~celti/esoterica/journalctl-list-boots-hang-redux.txt.xz

**Comment by Celti** [comment]
After restoring the backup and experimenting with `journalctl --vacuum-files` in single-file increments, journalctl --list-boots switches from the mmap/munmap cycle at 29 or more journal files present (exact files do not matter, experimented with removing random ones — only number seems to matter) to the incrementing brks at 28 or fewer files, to working as expected at 20 or fewer journal files.

**Comment by poettering** [comment]
Fix waiting in #3114.
